consensus/src/processes/pruning_proof/mod.rs

```diff
@@ -614,45 +714,285 @@ impl PruningProofManager {
 
         drop(pruning_read);
         drop(relations_read);
-        drop(db_lifetime);
+        drop(stores_and_processes.db_lifetime);
 
         Err(PruningImportError::PruningProofNotEnoughHeaders)
     }
 
+    // TODO: Find a better name
+    fn find_current_dag_level(&self, pp_header: &Header) -> BlockLevel {
+        let direct_parents = BlockHashSet::from_iter(pp_header.direct_parents().iter().copied());
+        pp_header
+            .parents_by_level
+            .iter()
+            .enumerate()
+            .skip(1)
+            .find_map(|(level, parents)| {
+                if BlockHashSet::from_iter(parents.iter().copied()) == direct_parents {
+                    None
+                } else {
+                    Some((level - 1) as BlockLevel)
+                }
+            })
+            .unwrap_or(self.max_block_level)
+    }
+
+    fn estimated_blue_depth_at_level_0(&self, level: BlockLevel, level_depth: u64, current_dag_level: BlockLevel) -> u64 {
+        level_depth << current_dag_level.saturating_sub(level)
+    }
+
+    fn find_selected_parent_header_at_level(
+        &self,
+        header: &Header,
+        level: BlockLevel,
+    ) -> PruningProofManagerInternalResult<Arc<Header>> {
+        let parents = self.parents_manager.parents_at_level(header, level);
+        let mut sp = SortableBlock { hash: parents[0], blue_work: self.headers_store.get_blue_score(parents[0]).unwrap_or(0).into() };
+        for parent in parents.iter().copied().skip(1) {
+            let sblock = SortableBlock {
+                hash: parent,
+                blue_work: self
+                    .headers_store
+                    .get_blue_score(parent)
+                    .unwrap_option()
+                    .ok_or(PruningProofManagerInternalError::NotEnoughHeadersToBuildProof(format!(
+                        "find_selected_parent_header_at_level (level {level}) couldn't find the header for block {parent}"
+                    )))?
+                    .into(),
+            };
+            if sblock > sp {
+                sp = sblock;
+            }
+        }
+        // TODO: For higher levels the chance of having more than two parents is very small, so it might make sense to fetch the whole header for the SortableBlock instead of blue_score (which will probably come from a compact header).
+        self.headers_store.get_header(sp.hash).unwrap_option().ok_or(PruningProofManagerInternalError::NotEnoughHeadersToBuildProof(
+            format!("find_selected_parent_header_at_level (level {level}) couldn't find the header for block {}", sp.hash,),
+        ))
+        // Ok(self.headers_store.get_header(sp.hash).unwrap_option().expect("already checked if compact header exists above"))
+    }
+
+    fn find_sufficient_root(
```

Comment:
What constitutes a "sufficient root"?

Reply:
It's a block that is at least 2M depth from the selected tip at current level that is also in the selected_chain(block_at_depth_m_at_next_level)